src/gfn/gflownet/base.py

```diff
@@ -1,18 +1,23 @@
-from abc import abstractmethod
-from typing import Tuple
+from abc import ABC, abstractmethod
+from typing import Generic, Tuple, TypeVar, Union
 
 import torch
 import torch.nn as nn
 from torchtyping import TensorType as TT
 
 from gfn.containers import Trajectories
+from gfn.containers.base import Container
 from gfn.env import Env
 from gfn.modules import GFNModule
 from gfn.samplers import Sampler
 from gfn.states import States
 
+TrainingSampleType = TypeVar(
+    "TrainingSampleType", bound=Union[Container, tuple[States, ...]]
+)
 
-class GFlowNet(nn.Module):
+
+class GFlowNet(ABC, nn.Module, Generic[TrainingSampleType]):
```

Comment:
I think this addition requires us to be absolutely sure that all class inheritance is working properly. In many cases we don't specify the `super().__init__()` because `GFlowNet` only inherited from `nn.Module`. In particular we need to ensure that inheritance from `nn.Module` works properly such that things like ` DBGFlowNet.parameters()` still returns all trainable parameters. Did you verify that these changes don't interfere with core pytorch functionality? Perhaps we need more tests to verify this.

Reply:
Hey @josephdviviano, thanks so much for the feedback and the great point here. I've added a test which checks that both `TBGFlowNet` and `FMGFlowNet` have working methods from `nn.Module` directly accessible. These two classes represent the only two ways that current subclasses of `GFlowNet` come into existence:

- `TBGFlowNet` inherits from `TrajectoryBasedGFlowNet` (and calls `super.__init__`). `TrajectoryBasedGFlowNet` then inherits from `PFBasedGFlowNet` which inherits from `GFlowNet` and calls `super.__init__()`. So thus, while `TrajectoryBasedGFlowNet` doesn't `super.__init__()`, that should be fine because it also doesn't redefine `self.__init__`.
- `FMGFlowNet` inherits from `GFlowNet` directly and calls `super.__init__()`

The test current looks like:

```python
def test_pytorch_inheritance():
    pf_module = BoxPFNeuralNet(
        hidden_dim=32, n_hidden_layers=2, n_components=1, n_components_s0=1
    )
    pb_module = BoxPBNeuralNet(
        hidden_dim=32, n_hidden_layers=2, n_components=1, torso=pf_module.torso
    )

    env = Box()

    pf_estimator = BoxPFEstimator(
        env=env, module=pf_module, n_components=1, n_components_s0=1
    )
    pb_estimator = BoxPBEstimator(env=env, module=pb_module, n_components=1)

    tbgflownet = TBGFlowNet(pf=pf_estimator, pb=pb_estimator)
    assert hasattr(
        tbgflownet.parameters(), "__iter__"
    ), "Expected gflownet to have iterable parameters() method inherited from nn.Module"
    assert hasattr(
        tbgflownet.state_dict(), "__dict__"
    ), "Expected gflownet to have indexable state_dict() method inherited from nn.Module"

    estimator = DiscretePolicyEstimator(env, pf_module, True)
    fmgflownet = FMGFlowNet(estimator)
    assert hasattr(
        fmgflownet.parameters(), "__iter__"
    ), "Expected gflownet to have iterable parameters() method inherited from nn.Module"
    assert hasattr(
        fmgflownet.state_dict(), "__dict__"
    ), "Expected gflownet to have indexable state_dict() method inherited from nn.Module"
```

It would be interesting to consider trying to make this more generic so that it can be more cleanly extended to test other kinds of `GFlowNet` subclasses. In Go, there is a cool pattern called [table tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) which would work well here. It's a bit more complicated to do this cleanly here though because each instance of a `GFlowNet` subclass often requires an overlapping set of dependencies so isolating them into their own scope would probably introduce additional code duplication. Happy to iterate though depending on what y'all think is best.

Reply:
Yes, exactly! I think the fact that it wasn't there must have been an oversight so I added it back in there.